### ISSUE TITLE
Memory Leak in Injector

### ISSUE THREAD

**Issue by GeorgianStan** [angular/components#22292]
I've encountered a possible memory leak while using `OverlayRef` and `Injector`.

#### Reproduction

Inside `AppComponent` there is a click event on a button. That click event will generate an array and pass that array to another component using `ComponentPortal`

The constructor and the click event from `AppComponent`:

```
// ? AppComponent constructor
  constructor(private overlay: Overlay) {
    this.#overlayRef = this.overlay.create({
      positionStrategy: this.overlay
        .position()
        .global()
        .centerHorizontally()
        .centerVertically(),
      hasBackdrop: true,
    });

    this.#overlayRef.backdropClick().subscribe(() => this.#overlayRef.detach());
}

onClick(){
 const output =  new Array(1000).fill('asfasdf');
 this.#overlayRef.attach(
      new ComponentPortal(
        MyComponent,
        null,
        Injector.create({
          providers: [{ provide: MY_TOKEN, useValue: output }],
        })
      )
    );
}
```
 

#### Actual Behavior

It looks like each time the modal is created the data generated and injected in the component, but once the modal disappears, the data is not removed from the memory. 

![1](https://user-images.githubusercontent.com/25796932/111782318-8c5e7b80-88c1-11eb-9797-ef426b18963d.PNG)
![2](https://user-images.githubusercontent.com/25796932/111782313-8b2d4e80-88c1-11eb-96a1-2790a5506f16.PNG)
![3](https://user-images.githubusercontent.com/25796932/111782316-8bc5e500-88c1-11eb-9976-c9184fad91f3.PNG)

#### Environment

- Angular:
  -  "@angular/core": "~11.2.5",
  
- CDK/Material:
  - "@angular/cdk": "^11.2.4",
- Browser(s): Chrome
- Operating System (e.g. Windows, macOS, Ubuntu): Windows10

**Comment by YoeriNijs** [comment]
I am not sure, but I guess you should call dispose() after detach() in order to actually clear the data entirely. Have you tested this?

**Comment by YoeriNijs** [comment]
I've added a reproduction in Spectator. See the link above.
